pkg_resources/__init__.py

```diff
@@ -2667,6 +2667,19 @@ def __getattr__(self, attr):
             raise AttributeError(attr)
         return getattr(self._provider, attr)
 
+    def __dir__(self):
+        return list(
+            set(super(Distribution, self).__dir__())
+            | set(
+                attr for attr in self._provider.__dir__()
+                if not attr.startswith('_')
+            )
+        )
+
+    if not hasattr(object, '__dir__'):
+        # python 2.7 not supported
+        del __dir__
```

Comment:
Python 2.7 doesn't provide a default implementation of `__dir__`, but it'll use one in `dir()` if present.  I found a pretty thorough [cross-version solution](https://stackoverflow.com/a/32453194) on StackOverflow for making sure a functional base `__dir__` is inherited; you'd probably only need a subset of that implementation to handle this particular case.

Reply:
I'd be inclined to adopt that implementation for backward compatibility if it were published in six. Then I'd trust that it's hardened and supported. Otherwise, I'd rather not adopt that code for the marginal benefit it would provide. Still, good reference.